File: cpp/endpoints/autonomy/execute_task_srv.cpp
#include "rclcpp/rclcpp.hpp"
#include "clearpath_task_msgs/srv/execute_task.hpp"

#include <chrono>
#include <cstdlib>
#include <memory>
#include <yaml-cpp/yaml.h>

using namespace std::chrono_literals;

// TODO: Enter your task id
const std::string TASK_ID = "";

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  
  YAML::Node robot_config = YAML::LoadFile("/etc/clearpath/robot.yaml");
  std::string namespace_ = robot_config["system"]["ros2"]["namespace"].as<std::string>();

  std::shared_ptr<rclcpp::Node> node = rclcpp::Node::make_shared("execute_task_srv_client");
  rclcpp::Client<clearpath_task_msgs::srv::ExecuteTask>::SharedPtr client =
    node->create_client<clearpath_task_msgs::srv::ExecuteTask>("/" + namespace_ + "/" + std::string("execute_task"));

  auto request = std::make_shared<clearpath_task_msgs::srv::ExecuteTask::Request>();
  request->task_id = TASK_ID;

  while (!client->wait_for_service(2s)) {
    if (!rclcpp::ok()) {
      RCLCPP_ERROR(rclcpp::get_logger("execute_task_srv_client"), "Interrupted while waiting for the service. Exiting.");
      return 0;
    }
    RCLCPP_INFO(rclcpp::get_logger("execute_task_srv_client"), "/%s/execute_task service not available, waiting again...", namespace_.c_str());
  }

  auto result = client->async_send_request(request);

  // Wait for the result.
  if (rclcpp::spin_until_future_complete(node, result) ==
    rclcpp::FutureReturnCode::SUCCESS)
  {
    if (result.get()->success)
    {
      RCLCPP_INFO(rclcpp::get_logger("execute_task_srv_client"), "Task executed successfully!");
    }
    else
    {
      RCLCPP_ERROR(rclcpp::get_logger("execute_task_srv_client"), "Failed to execute task!");
    }
  } else {
    RCLCPP_ERROR(rclcpp::get_logger("execute_task_srv_client"), "Failed to call service /%s/execute_task", namespace_.c_str());
  }

  rclcpp::shutdown();
  return 0;
}

File: python/endpoints/autonomy/execute_task_srv.py
from clearpath_config.common.utils.yaml import read_yaml
from clearpath_task_msgs.srv import ExecuteTask
import rclpy
from rclpy.node import Node

ROBOT_CONFIG_PATH = '/etc/clearpath/robot.yaml'
LOGGING_NAME = 'ExecuteTaskSrv'

# TODO: Enter your task uuid
TASK_ID: str = ''


class ExecuteTaskSrv(Node):

    def __init__(self, namespace: str):
        super().__init__('execute_task_srv')
        self._execute_task_srv_client = self.create_client(ExecuteTask, f'/{namespace}/autonomy/task/execute_srv')

        while not self._execute_task_srv_client.wait_for_service(timeout_sec=2.0):
            self.get_logger().info(f'[{LOGGING_NAME}] /{namespace}/autonomy/task/execute_srv service not available, waiting again...')

    def execute_task(self):
        req = ExecuteTask.Request()
        req.task_id = TASK_ID
        self.future = self._execute_task_srv_client.call_async(req)
        self.future.add_done_callback(self.execute_task_srv_response_cb)

    def execute_task_srv_response_cb(self, future):
        try:
            response = future.result()
            if response.success:
                self.get_logger().info(f'[{LOGGING_NAME}] Task executed successfully!')
            else:
                self.get_logger().error(f'[{LOGGING_NAME}] Failed to execute task')
        except Exception as e:
            self.get_logger().error(f'[{LOGGING_NAME}] Service call failed: {e}')

        self.destroy_node()
        rclpy.shutdown()


def main(args=None):
    rclpy.init(args=args)
    robot_config = read_yaml(ROBOT_CONFIG_PATH)
    namespace = robot_config['system']['ros2']['namespace']

    execute_task_client = ExecuteTaskSrv(namespace)
    execute_task_client.execute_task()

    rclpy.spin(execute_task_client)


if __name__ == '__main__':
    main()
